diff --git a/tpm2/constants.go b/tpm2/constants.go
index 70edae2..0f8b740 100644
--- a/tpm2/constants.go
+++ b/tpm2/constants.go
@@ -75,6 +75,27 @@ const (
 	TPMAlgKMACXOF256   TPMAlgID = 0x0082
 	TPMAlgKMAC128      TPMAlgID = 0x0090
 	TPMAlgKMAC256      TPMAlgID = 0x0091
+
+	// 6.3 TPM_ALG_ID pg 41 https://trustedcomputinggroup.org/wp-content/uploads/Trusted-Platform-Module-2.0-Library-Part-2-Structures_Version-185_pub.pdf
+	TPMAlgMLKEM     TPMAlgID = 0x00A0
+	TPMAlgMLDSA     TPMAlgID = 0x00A1
+	TPMAlgHASHMLDSA TPMAlgID = 0x00A2
+)
+
+type TPMMLKEMParameter uint16
+
+const (
+	TPMMLKEM512  TPMMLKEMParameter = 0x0001
+	TPMMLKEM768  TPMMLKEMParameter = 0x0002
+	TPMMLKEM1024 TPMMLKEMParameter = 0x0003
+)
+
+type TPMMLDSAParameter uint16
+
+const (
+	TPMMLDSA44 TPMMLDSAParameter = 0x0001
+	TPMMLDSA65 TPMMLDSAParameter = 0x0002
+	TPMMLDSA87 TPMMLDSAParameter = 0x0003
 )
 
 // TPMECCCurve represents a TPM_ECC_Curve.
@@ -222,6 +243,16 @@ const (
 	TPMCCPolicyACSendSelect         TPMCC = 0x00000196
 	TPMCCCertifyX509                TPMCC = 0x00000197
 	TPMCCACTSetTimeout              TPMCC = 0x00000198
+
+	// pg 52 https://trustedcomputinggroup.org/wp-content/uploads/Trusted-Platform-Module-2.0-Library-Part-2-Structures_Version-185_pub.pdf
+	TPMCCEncapsulate            TPMCC = 0x000001A7
+	TPMCCDecapsulate            TPMCC = 0x000001A8
+	TPMCCSignSequenceStart      TPMCC = 0x000001AA
+	TPMCCVerifySequenceStart    TPMCC = 0x000001A9
+	TPMCCSignSequenceComplete   TPMCC = 0x000001A4
+	TPMCCVerifySequenceComplete TPMCC = 0x000001A3
+	TPMCCSignDigest             TPMCC = 0x000001A6
+	TPMCCVerifyDigestSignature  TPMCC = 0x000001A5
 )
 
 // TPMRC represents a TPM_RC.
diff --git a/tpm2/structures.go b/tpm2/structures.go
index 17a6dd9..c92457f 100644
--- a/tpm2/structures.go
+++ b/tpm2/structures.go
@@ -1921,6 +1921,11 @@ func (u *TPMUSigScheme) create(hint int64) (reflect.Value, error) {
 		u.contents = &contents
 		u.selector = TPMAlgID(hint)
 		return reflect.ValueOf(&contents), nil
+	case TPMAlgMLDSA:
+		var contents TPM2BData
+		u.contents = &contents
+		u.selector = TPMAlgID(hint)
+		return reflect.ValueOf(&contents), nil
 	}
 	return reflect.ValueOf(nil), fmt.Errorf("no union member for tag %v", hint)
 }
@@ -1949,6 +1954,12 @@ func (u TPMUSigScheme) get(hint int64) (reflect.Value, error) {
 			contents = *u.contents.(*TPMSSchemeECDAA)
 		}
 		return reflect.ValueOf(&contents), nil
+	case TPMAlgMLDSA:
+		var contents TPM2BData
+		if u.contents != nil {
+			contents = *u.contents.(*TPM2BData)
+		}
+		return reflect.ValueOf(&contents), nil
 	}
 	return reflect.ValueOf(nil), fmt.Errorf("no union member for tag %v", hint)
 }
@@ -2417,6 +2428,9 @@ type TPM2BPrivateKeyRSA TPM2BData
 // See definition in Part 2: Structures, section 11.2.5.1.
 type TPM2BECCParameter TPM2BData
 
+type TPM2BPublicKeyMLKEM TPM2BData
+type TPM2BSignatureContext TPM2BData
+
 // TPMSECCPoint represents a TPMS_ECC_POINT.
 // See definition in Part 2: Structures, section 11.2.5.2.
 type TPMSECCPoint struct {
@@ -2476,10 +2490,15 @@ type TPMUSignature struct {
 	contents Marshallable
 }
 
+type TPMSSigSchemeMLDSA struct {
+	marshalByReflection
+	Signature TPM2BData
+}
+
 // SignatureContents is a type constraint representing the possible contents of TPMUSignature.
 type SignatureContents interface {
 	Marshallable
-	*TPMTHA | *TPMSSignatureRSA | *TPMSSignatureECC
+	*TPMTHA | *TPMSSignatureRSA | *TPMSSignatureECC | *TPMSSigSchemeMLDSA
 }
 
 // create implements the unmarshallableWithHint interface.
@@ -2500,6 +2519,12 @@ func (u *TPMUSignature) create(hint int64) (reflect.Value, error) {
 		u.contents = &contents
 		u.selector = TPMAlgID(hint)
 		return reflect.ValueOf(&contents), nil
+	case TPMAlgMLDSA:
+		var contents TPMSSigSchemeMLDSA
+		u.contents = &contents
+		u.selector = TPMAlgID(hint)
+		return reflect.ValueOf(&contents), nil
+
 	}
 	return reflect.ValueOf(nil), fmt.Errorf("no union member for tag %v", hint)
 }
@@ -2528,6 +2553,12 @@ func (u TPMUSignature) get(hint int64) (reflect.Value, error) {
 			contents = *u.contents.(*TPMSSignatureECC)
 		}
 		return reflect.ValueOf(&contents), nil
+	case TPMAlgMLDSA:
+		var contents TPMSSigSchemeMLDSA
+		if u.contents != nil {
+			contents = *u.contents.(*TPMSSigSchemeMLDSA)
+		}
+		return reflect.ValueOf(&contents), nil
 	}
 	return reflect.ValueOf(nil), fmt.Errorf("no union member for tag %v", hint)
 }
@@ -2580,6 +2611,13 @@ func (u *TPMUSignature) ECDAA() (*TPMSSignatureECC, error) {
 	return nil, fmt.Errorf("did not contain ecdaa (selector value was %v)", u.selector)
 }
 
+func (u *TPMUSignature) MLDSA() (*TPMSSigSchemeMLDSA, error) {
+	if u.selector == TPMAlgMLDSA {
+		return u.contents.(*TPMSSigSchemeMLDSA), nil
+	}
+	return nil, fmt.Errorf("did not contain ecdaa (selector value was %v)", u.selector)
+}
+
 // TPMTSignature represents a TPMT_SIGNATURE.
 // See definition in Part 2: Structures, section 11.3.4.
 type TPMTSignature struct {
@@ -2608,7 +2646,7 @@ type TPMUPublicID struct {
 // PublicIDContents is a type constraint representing the possible contents of TPMUPublicID.
 type PublicIDContents interface {
 	Marshallable
-	*TPM2BDigest | *TPM2BPublicKeyRSA | *TPMSECCPoint
+	*TPM2BDigest | *TPM2BPublicKeyRSA | *TPMSECCPoint | *TPM2BData
 }
 
 // create implements the unmarshallableWithHint interface.
@@ -2634,6 +2672,16 @@ func (u *TPMUPublicID) create(hint int64) (reflect.Value, error) {
 		u.contents = &contents
 		u.selector = TPMAlgID(hint)
 		return reflect.ValueOf(&contents), nil
+	case TPMAlgMLKEM:
+		var contents TPM2BData
+		u.contents = &contents
+		u.selector = TPMAlgID(hint)
+		return reflect.ValueOf(&contents), nil
+	case TPMAlgMLDSA:
+		var contents TPM2BData
+		u.contents = &contents
+		u.selector = TPMAlgID(hint)
+		return reflect.ValueOf(&contents), nil
 	}
 	return reflect.ValueOf(nil), fmt.Errorf("no union member for tag %v", hint)
 }
@@ -2668,6 +2716,18 @@ func (u TPMUPublicID) get(hint int64) (reflect.Value, error) {
 			contents = *u.contents.(*TPMSECCPoint)
 		}
 		return reflect.ValueOf(&contents), nil
+	case TPMAlgMLKEM:
+		var contents TPMSMLKEMParms
+		if u.contents != nil {
+			contents = *u.contents.(*TPMSMLKEMParms)
+		}
+		return reflect.ValueOf(&contents), nil
+	case TPMAlgMLDSA:
+		var contents TPMSMLDSAParms
+		if u.contents != nil {
+			contents = *u.contents.(*TPMSMLDSAParms)
+		}
+		return reflect.ValueOf(&contents), nil
 	}
 	return reflect.ValueOf(nil), fmt.Errorf("no union member for tag %v", hint)
 }
@@ -2712,6 +2772,23 @@ func (u *TPMUPublicID) ECC() (*TPMSECCPoint, error) {
 	return nil, fmt.Errorf("did not contain ecc (selector value was %v)", u.selector)
 }
 
+func (u *TPMUPublicID) KEM() (*TPM2BData, error) {
+	if u.selector == TPMAlgMLKEM {
+		return u.contents.(*TPM2BData), nil
+	}
+	return nil, fmt.Errorf("did not contain TPMAlgMLKEM (selector value was %v)", u.selector)
+}
+
+func (u *TPMUPublicID) MLDSA() (*TPM2BData, error) {
+	if u.selector == TPMAlgMLDSA {
+		return u.contents.(*TPM2BData), nil
+	}
+	return nil, fmt.Errorf("did not contain TPMAlgMLDSA (selector value was %v)", u.selector)
+}
+
+type TPM2BSharedSecret TPM2BData
+type TPM2BKEMCipherText TPM2BData
+
 // TPMSKeyedHashParms represents a TPMS_KEYEDHASH_PARMS.
 // See definition in Part 2: Structures, section 12.2.3.3.
 type TPMSKeyedHashParms struct {
@@ -2747,6 +2824,21 @@ type TPMSRSAParms struct {
 	// A prime number greater than 2.
 	Exponent uint32
 }
+type TPMIMLDSAParam = uint16
+type TPMSMLDSAParms struct {
+	marshalByReflection
+	ParameterSet    TPMIMLDSAParam
+	AllowExternalMu TPMIYesNo
+}
+
+// 11.2.6 ML-KEM pg 182 https://trustedcomputinggroup.org/wp-content/uploads/Trusted-Platform-Module-2.0-Library-Part-2-Structures_Version-185_pub.pdf
+
+type TPMIMLKEMParam = uint16
+type TPMSMLKEMParms struct {
+	marshalByReflection
+	Symmetric    TPMTSymDefObject
+	ParameterSet TPMIMLKEMParam
+}
 
 // TPMSECCParms represents a TPMS_ECC_PARMS.
 // See definition in Part 2: Structures, section 12.2.3.6.
@@ -2780,7 +2872,7 @@ type TPMUPublicParms struct {
 type PublicParmsContents interface {
 	Marshallable
 	*TPMSKeyedHashParms | *TPMSSymCipherParms | *TPMSRSAParms |
-		*TPMSECCParms
+		*TPMSECCParms | *TPMSMLKEMParms | *TPMSMLDSAParms
 }
 
 // create implements the unmarshallableWithHint interface.
@@ -2806,6 +2898,16 @@ func (u *TPMUPublicParms) create(hint int64) (reflect.Value, error) {
 		u.contents = &contents
 		u.selector = TPMAlgID(hint)
 		return reflect.ValueOf(&contents), nil
+	case TPMAlgMLKEM:
+		var contents TPMSMLKEMParms
+		u.contents = &contents
+		u.selector = TPMAlgID(hint)
+		return reflect.ValueOf(&contents), nil
+	case TPMAlgMLDSA:
+		var contents TPMSMLDSAParms
+		u.contents = &contents
+		u.selector = TPMAlgID(hint)
+		return reflect.ValueOf(&contents), nil
 	}
 	return reflect.ValueOf(nil), fmt.Errorf("no union member for tag %v", hint)
 }
@@ -2840,6 +2942,18 @@ func (u TPMUPublicParms) get(hint int64) (reflect.Value, error) {
 			contents = *u.contents.(*TPMSECCParms)
 		}
 		return reflect.ValueOf(&contents), nil
+	case TPMAlgMLKEM:
+		var contents TPMSMLKEMParms
+		if u.contents != nil {
+			contents = *u.contents.(*TPMSMLKEMParms)
+		}
+		return reflect.ValueOf(&contents), nil
+	case TPMAlgMLDSA:
+		var contents TPMSMLDSAParms
+		if u.contents != nil {
+			contents = *u.contents.(*TPMSMLDSAParms)
+		}
+		return reflect.ValueOf(&contents), nil
 	}
 	return reflect.ValueOf(nil), fmt.Errorf("no union member for tag %v", hint)
 }
@@ -2884,6 +2998,20 @@ func (u *TPMUPublicParms) ECCDetail() (*TPMSECCParms, error) {
 	return nil, fmt.Errorf("did not contain eccDetail (selector value was %v)", u.selector)
 }
 
+func (u *TPMUPublicParms) MLKEMDetail() (*TPMSMLKEMParms, error) {
+	if u.selector == TPMAlgMLKEM {
+		return u.contents.(*TPMSMLKEMParms), nil
+	}
+	return nil, fmt.Errorf("did not contain mlkem (selector value was %v)", u.selector)
+}
+
+func (u *TPMUPublicParms) MLDSADetail() (*TPMSMLDSAParms, error) {
+	if u.selector == TPMAlgMLDSA {
+		return u.contents.(*TPMSMLDSAParms), nil
+	}
+	return nil, fmt.Errorf("did not contain mldsa (selector value was %v)", u.selector)
+}
+
 // TPMTPublicParms represents a TPMT_PUBLIC_PARMS.
 // See definition in Part 2: Structures, section 12.2.3.8.
 type TPMTPublicParms struct {
diff --git a/tpm2/tpm2.go b/tpm2/tpm2.go
index e921c0b..bf570ae 100644
--- a/tpm2/tpm2.go
+++ b/tpm2/tpm2.go
@@ -268,6 +268,52 @@ type LoadResponse struct {
 	Name TPM2BName
 }
 
+type Encapsulate struct {
+	// reference to public portion of symmetric key to use for encryption
+	KeyHandle handle `gotpm:"handle,auth"`
+}
+
+// Command implements the Command interface.
+func (Encapsulate) Command() TPMCC { return TPMCCEncapsulate }
+
+// Execute executes the command and returns the response.
+func (cmd Encapsulate) Execute(t transport.TPM, s ...Session) (*EncapsulateResponse, error) {
+	var rsp EncapsulateResponse
+	err := execute[EncapsulateResponse](t, cmd, &rsp, s...)
+	if err != nil {
+		return nil, err
+	}
+	return &rsp, nil
+}
+
+type EncapsulateResponse struct {
+	SharedSecret TPM2BSharedSecret
+	CipherText   TPM2BKEMCipherText
+}
+
+type Decapsulate struct {
+	// reference to public portion of symmetric key to use for encryption
+	KeyHandle  handle `gotpm:"handle,auth"`
+	CipherText TPM2BKEMCipherText
+}
+
+// Command implements the Command interface.
+func (Decapsulate) Command() TPMCC { return TPMCCDecapsulate }
+
+// Execute executes the command and returns the response.
+func (cmd Decapsulate) Execute(t transport.TPM, s ...Session) (*DecapsulateResponse, error) {
+	var rsp DecapsulateResponse
+	err := execute[DecapsulateResponse](t, cmd, &rsp, s...)
+	if err != nil {
+		return nil, err
+	}
+	return &rsp, nil
+}
+
+type DecapsulateResponse struct {
+	SharedSecret TPM2BSharedSecret
+}
+
 // LoadExternal is the input to TPM2_LoadExternal.
 // See definition in Part 3, Commands, section 12.3
 type LoadExternal struct {
@@ -1034,6 +1080,131 @@ type SignResponse struct {
 	Signature TPMTSignature
 }
 
+type SignDigest struct {
+	// Handle of key that will perform signing
+	KeyHandle handle `gotpm:"handle,auth"`
+	// digest to be signed
+	Context    TPM2BSignatureContext
+	Digest     TPM2BDigest
+	Validation TPMTTKHashCheck
+}
+
+// Command implements the Command interface.
+func (SignDigest) Command() TPMCC { return TPMCCSignDigest }
+
+// Execute executes the command and returns the response.
+func (cmd SignDigest) Execute(t transport.TPM, s ...Session) (*SignDigestResponse, error) {
+	var rsp SignDigestResponse
+	if err := execute[SignDigestResponse](t, cmd, &rsp, s...); err != nil {
+		return nil, err
+	}
+	return &rsp, nil
+}
+
+// SignResponse is the response from TPM2_Sign.
+type SignDigestResponse struct {
+	// the signature
+	Signature TPMTSignature
+}
+
+type SignSequenceStart struct {
+	// Handle of key that will perform signing
+	KeyHandle handle `gotpm:"handle,auth"`
+	// authorization value for subsequent use of the sequence
+	Auth    TPM2BAuth
+	Context TPM2BSignatureContext
+}
+
+// Command implements the Command interface.
+func (SignSequenceStart) Command() TPMCC { return TPMCCSignSequenceStart }
+
+// Execute executes the command and returns the response.
+func (cmd SignSequenceStart) Execute(t transport.TPM, s ...Session) (*SignSequenceStartResponse, error) {
+	var rsp SignSequenceStartResponse
+	if err := execute[SignSequenceStartResponse](t, cmd, &rsp, s...); err != nil {
+		return nil, err
+	}
+	return &rsp, nil
+}
+
+// SignResponse is the response from TPM2_Sign.
+type SignSequenceStartResponse struct {
+	// the signature
+	SequenceHandle TPMIDHObject `gotpm:"handle"`
+}
+
+type SignSequenceComplete struct {
+	SequenceHandle handle `gotpm:"handle,auth,anon"`
+	KeyHandle      handle `gotpm:"handle,auth"`
+	// data to be added to the hash/HMAC
+	Buffer TPM2BMaxBuffer
+}
+
+// Command implements the Command interface.
+func (SignSequenceComplete) Command() TPMCC { return TPMCCSignSequenceComplete }
+
+// Execute executes the command and returns the response.
+func (cmd SignSequenceComplete) Execute(t transport.TPM, s ...Session) (*SignSequenceCompleteResponse, error) {
+	var rsp SignSequenceCompleteResponse
+	if err := execute[SignSequenceCompleteResponse](t, cmd, &rsp, s...); err != nil {
+		return nil, err
+	}
+	return &rsp, nil
+}
+
+// SignResponse is the response from TPM2_Sign.
+type SignSequenceCompleteResponse struct {
+	Signature TPMTSignature
+}
+
+type VerifySequenceStart struct {
+	KeyHandle handle `gotpm:"handle,auth"`
+	Auth      TPM2BAuth
+	Hint      TPM2BData
+	Context   TPM2BSignatureContext
+}
+
+// Command implements the Command interface.
+func (VerifySequenceStart) Command() TPMCC { return TPMCCVerifySequenceStart }
+
+// Execute executes the command and returns the response.
+func (cmd VerifySequenceStart) Execute(t transport.TPM, s ...Session) (*VerifySequenceStartResponse, error) {
+	var rsp VerifySequenceStartResponse
+	if err := execute[VerifySequenceStartResponse](t, cmd, &rsp, s...); err != nil {
+		return nil, err
+	}
+	return &rsp, nil
+}
+
+// SignResponse is the response from TPM2_Sign.
+type VerifySequenceStartResponse struct {
+	// the signature
+	SequenceHandle TPMIDHObject `gotpm:"handle"`
+}
+
+type VerifySequenceComplete struct {
+	SequenceHandle handle `gotpm:"handle,auth,anon"`
+	KeyHandle      handle `gotpm:"handle,auth"`
+	Signature      TPMTSignature
+}
+
+// Command implements the Command interface.
+func (VerifySequenceComplete) Command() TPMCC { return TPMCCVerifySequenceComplete }
+
+// Execute executes the command and returns the response.
+func (cmd VerifySequenceComplete) Execute(t transport.TPM, s ...Session) (*VerifySequenceCompleteResponse, error) {
+	var rsp VerifySequenceCompleteResponse
+	if err := execute[VerifySequenceCompleteResponse](t, cmd, &rsp, s...); err != nil {
+		return nil, err
+	}
+	return &rsp, nil
+}
+
+// SignResponse is the response from TPM2_Sign.
+type VerifySequenceCompleteResponse struct {
+	Validation TPMTTKVerified
+}
+
 // PCRExtend is the input to TPM2_PCR_Extend.
 // See definition in Part 3, Commands, section 22.2
 type PCRExtend struct {
